Use thrust to generate the coarse of amgx_pgm

yhmtsai opened the pull request:
This pull request eliminate the usage of spgemm in coarse grid generation of amgx_pgm
After pull request, the transposed is only needed by the weight matrix computation.
If the matrix is hermitian, the transpose can also be ignored.

upsj commented on cuda/multigrid/amgx_pgm_kernels.cu:
```diff
@@ -0,0 +1,133 @@
+        thrust::device_pointer_cast(coarse_coo->get_row_idxs()),
+        thrust::device_pointer_cast(coarse_coo->get_col_idxs())));
+
+    thrust::reduce_by_key(thrust::device, key_it, key_it + fine_nnz, vals_it,
```
note: reduce_by_key can be non-deterministic

upsj commented on cuda/multigrid/amgx_pgm_kernels.cu:
```diff
@@ -0,0 +1,133 @@
+    auto it = thrust::make_zip_iterator(
+        thrust::make_tuple(thrust::device_pointer_cast(row_idxs),
+                           thrust::device_pointer_cast(col_idxs)));
+    thrust::stable_sort_by_key(thrust::device, it, it + nnz, vals_it);
```
nit: does this need to be stable?
```suggestion
    thrust::sort_by_key(thrust::device, it, it + nnz, vals_it);
```

yhmtsai replied:
Yes only when reduce_by_key is determinstic. If reduce_by_key is not determinstic, this stable version does not help the coarse determinstic property

upsj commented on dpcpp/multigrid/amgx_pgm_kernels.dp.cpp:
```diff
@@ -0,0 +1,132 @@
+    auto policy =
+        oneapi::dpl::execution::make_device_policy(*exec->get_queue());
+    auto it = oneapi::dpl::make_zip_iterator(row_idxs, col_idxs, vals);
+    std::stable_sort(policy, it, it + nnz, [](auto a, auto b) {
```
same here
```suggestion
    std::sort(policy, it, it + nnz, [](auto a, auto b) {
```

upsj commented on hip/multigrid/amgx_pgm_kernels.hip.cpp:
```diff
@@ -0,0 +1,131 @@
+
+
+template <typename IndexType>
+void sort_agg(std::shared_ptr<const DefaultExecutor> exec, IndexType num,
```
can be pulled out into common/cuda_hip

upsj commented on omp/multigrid/amgx_pgm_kernels.cpp:
```diff
@@ -0,0 +1,129 @@
+                    IndexType* row_idxs, IndexType* col_idxs, ValueType* vals)
+{
+    auto it = detail::make_zip_iterator(row_idxs, col_idxs, vals);
+    std::stable_sort(it, it + nnz, [](auto a, auto b) {
```
```suggestion
    std::sort(it, it + nnz, [](auto a, auto b) {
```

upsj commented on test/matrix/matrix.cpp:
```diff
@@ -39,6 +39,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <gtest/gtest.h>
 
 
+#include <ginkgo/core/base/device_matrix_data.hpp>
```
is it possible the file isn't picked up by the formatting script?

upsj replied:
Has been fixed in #1011

upsj reviewed (approved):
LGTM! Minor nits. Also, this turns the previously deterministic generation (depending on SpGEMM determinism) into a non-deterministic one

pratikvn commented on hip/multigrid/amgx_pgm_kernels.hip.cpp:
```diff
@@ -0,0 +1,131 @@
+void sort_row_major(std::shared_ptr<const DefaultExecutor> exec, size_type nnz,
+                    IndexType* row_idxs, IndexType* col_idxs, ValueType* vals)
+{
+    // workaround for CUDA 9.2 Thrust: Their complex<> implementation is broken
```
Is this comment relevant for HIP thrust as well ?

upsj replied:
The comment was probably copied from my code: Yes, if you compile HIP for NVIDIA, you use NVIDIA Thrust, and rocThrust was forked from NVIDIA Thrust, so it might be that this issue is also present in an older rocThrust version.

pratikvn commented on omp/multigrid/amgx_pgm_kernels.cpp:
```diff
@@ -0,0 +1,129 @@
+    auto coarse_row = coarse_coo->get_row_idxs();
+    auto coarse_col = coarse_coo->get_col_idxs();
+    auto coarse_val = coarse_coo->get_values();
+    IndexType row = 0;
+    size_type idxs = 0;
```
```suggestion
```

yhmtsai replied:
I delete row only because I still use idx

pratikvn commented on reference/multigrid/amgx_pgm_kernels.cpp:
```diff
@@ -231,6 +296,59 @@ GKO_INSTANTIATE_FOR_EACH_NON_COMPLEX_VALUE_AND_INDEX_TYPE(
+                    IndexType* row_idxs, IndexType* col_idxs, ValueType* vals)
+{
+    auto it = detail::make_zip_iterator(row_idxs, col_idxs, vals);
+    std::stable_sort(it, it + nnz, [](auto a, auto b) {
```
Is there a particular reason you need `stable_sort` ?

yhmtsai replied:
when the reduce is determinstic, it will also require the element position determinstic.

pratikvn replied:
In each row, each value has a unique col_idx right ? If the the col_idx is the same, you sum them up later anyway, right ?

yhmtsai replied:
no, some value might have the same col_idx. it is used after fine matrix -> coarse matrix index mapping.

pratikvn reviewed (approved):
Really nice work! Do you have some performance numbers as to how much this improves the generation time over the previous approach ?

yhmtsai commented:
rebase!

yhmtsai commented:
@pratikvn there are some result. The performance is not so good, and the generation should be improved more in the future. This version is mainly to avoid the SpGEMM calls such that we can implement it in distributed version more easily.
L-shape
original: 82 80 80
updated: 65 65 64
beam
original: 80 80 74
updated: 74 72 72

pratikvn commented:
@yhmtsai, thanks for checking. I think that is still a significant improvement.
